### ISSUE TITLE
Feat: Like feature is not implemented properly.

### ISSUE THREAD

**Issue by Sanchitbajaj02** [Sanchitbajaj02/palettegram#92]
### What feature?

The code for like feature is implemented but it is not working properly on the server. A PoC is required before going further.

### Record

- [X] I agree to follow this project's Code of Conduct
- [ ] I want to work on this issue

**Comment by Archit5655** [comment]
I would like to take on this issue

**Comment by sumanjeet0012** [comment]
I want to work on this issue.
assign me

**Comment by Sanchitbajaj02** [comment]
@sumanjeet0012 assigned to you

**Comment by Jenis07** [comment]
Can you assign me this issue under IWOC?

**Comment by Sanchitbajaj02** [comment]
@Jenis07 sure. Assigned this to you. Make sure to update your progress and if you have any doubt, just let me know.

**Comment by IrfanshaikH-7** [comment]
Can I look into it ...@Sanchitbajaj02

**Comment by Manishak798** [comment]
@Jenis07 have you started working on this ?

**Comment by Sanchitbajaj02** [comment]
@Jenis07 If you don't respond within 24 hours, I will unassign the issue

**Comment by Jenis07** [comment]
Yes sir I have started working on this and I have made some logic for
that.I will look this by today.

**Comment by Sanchitbajaj02** [comment]
@Jenis07 any updates?

**Comment by utkarsh4321** [comment]
Hi @Sanchitbajaj02  can I work on this issue? But need to understand your requirements first.

**Comment by Sanchitbajaj02** [comment]
> Hi @Sanchitbajaj02 can I work on this issue? But need to understand your requirements first.

Sure assigned the issue to you. The requirement is to implement a simple like feature. When use click on the like button a count get updated into the post table and an entry will be entered into the likes table.
The like table creates a new record which contains a postId and a userId. This process will be handled in the queue which can handle multiple users creating and retrieving entries

**Comment by utkarsh4321** [comment]
@Sanchitbajaj02  ok i will check and if anything needed will talk to you on this.

**Comment by ArtiGaund** [comment]
So, there are User, Post, and like collections. When the user clicks on the like button, =>  if like is already enabled (checking whether post id with user id is present in the Like collections), it will delete that entry from the like collection and decrease the count of post ids in Post collection. Else will increase the count of post IDs in Post Collection and add entry in Like collection with post IDs and user IDs.

**Comment by Sanchitbajaj02** [comment]
> So, there are User, Post, and like collections. When the user clicks on the like button, => if like is already enabled (checking whether post id with user id is present in the Like collections), it will delete that entry from the like collection and decrease the count of post ids in Post collection. Else will increase the count of post IDs in Post Collection and add entry in Like collection with post IDs and user IDs.

### Yes, that's the aim of this functionality

**Comment by ArtiGaund** [comment]
I will try to do it.

**Comment by ArtiGaund** [comment]
![Screenshot (307)](https://github.com/Sanchitbajaj02/palettegram/assets/38074085/49241638-40d5-4fec-a49e-727587e89b98)
Login feature is not working, yesterday it was working, when I created account after setting up the project, but today when I am trying to login its showing this error. Even after I have modified the password in the databse its showing this error. I am not able to see the post page.

**Comment by ArtiGaund** [comment]
whats the timelimit for the assignment?

**Comment by Sanchitbajaj02** [comment]
> whats the timelimit for the assignment?

There is no as such time limit, if you need time, you can take it. Just keep me updated with the progress. As for the login issue, what you can do is peek into the appwrite and see whether in the auth you are verified or not. If yes, go to db and see whether an entry is created with your username. Let me know here or on discord about your findings...

**Comment by ArtiGaund** [comment]
For login, I am registering and deleting accounts daily, as I have appwriteDB in my appwrite console. Through registration, I am able to get a login automatically and can view the posts and work on them.

**Comment by Sanchitbajaj02** [comment]
> For login, I am registering and deleting accounts daily, as I have appwriteDB in my appwrite console. Through registration, I am able to get a login automatically and can view the posts and work on them.

Actually, you don't need to delete daily. What you can do is find out what part is giving the error in the component.

**Comment by ArtiGaund** [comment]
I will try to find it.

**Comment by ArtiGaund** [comment]
Resolved you registeration and login problem?

**Comment by Sanchitbajaj02** [comment]
> Resolved you registeration and login problem.

Can you address the issue and your findings in #300

**Comment by ArtiGaund** [comment]
as the appwrite version is fixed in the project, I cannot  use upgrade version methods? bz they will not work.

**Comment by Sanchitbajaj02** [comment]
> as the appwrite version is fixed in the project, I cannot  use upgrade version methods? bz they will not work.

Does your implementation require it to be on an upgraded version?

**Comment by ArtiGaund** [comment]
> > as the appwrite version is fixed in the project, I cannot  use upgrade version methods? bz they will not work.
> 
> Does your implementation require it to be on an upgraded version?

For performing like and dislike, I need to check whether the user has liked the post or not. To check that entry, their is a query.and() method that checks multiple queries together. But Query.and() is in the upgraded version of AppWrite.  The assignment is not yet complete, but bugs are there. That's why I am asking; I will change that and use a different method to do it.

**Comment by Sanchitbajaj02** [comment]
> But Query.and() is in the upgraded version of AppWrite. The assignment is not yet complete, but bugs are there. That's why I am asking; I will change that and use a different method to do it.

There will be multiple changes like that login issue? if you want we can switch to higher variant but its just that we need to cater those changes

**Comment by ArtiGaund** [comment]
Login issue is resolved now?

**Comment by Sanchitbajaj02** [comment]
> Login issue is resolved now?

That's not the issue. If it was working fine on version 13 then it isn't a bug. Since it is not recommended to upgrade the version when I provided the lock functionality

**Comment by ArtiGaund** [comment]
I will complete the like functionality first, which was assigned to me. Then we will pull upstream and check the error for login and registration.

**Comment by ArtiGaund** [comment]
![Screenshot (332)](https://github.com/Sanchitbajaj02/palettegram/assets/38074085/055bb486-f463-4f29-99f9-a116f2063038)
![Screenshot (333)](https://github.com/Sanchitbajaj02/palettegram/assets/38074085/0bbcf284-e4ac-471c-b995-a3c0d46e7f15)
![Screenshot (334)](https://github.com/Sanchitbajaj02/palettegram/assets/38074085/a6b1af41-7512-44b2-b04c-7e5199510aa7)

**Comment by Sanchitbajaj02** [comment]
Check whether the proper type is declared for onLikeClick. If you face any issue, please reach out to me on discord

**Comment by vinayak-vohra** [comment]
@Sanchitbajaj02 may I work on this issue?

**Comment by Sanchitbajaj02** [comment]
> @Sanchitbajaj02 may I work on this issue?

yeah sure

**Comment by vinayak-vohra** [comment]
@Sanchitbajaj02 the posts collection has a likesCount attribute of integer type which would only tell the number of likes. 
There is no way of knowing whether the current user has liked the post or not, which is needed to implement the like and unlike functionality. There is also a likes collection which is never used. I've thought of a few ways to tackle this issue:

1. **Change likesCount(integer) to likedBy(array of userIds)**: 
    - Like count would be the size of likedBy array. 
    - For Like/Unlike we can check the array for current userId. If present, splice the array and update document **(unlike)**. Else push current userId into the array **(like)**.
    - In this approach, userIds of all users who liked the post would be sent.
 2. **Utilizing the likes collection**: 
    - Likes collection is a one-to-one relation between userId and postId, we can fetch all the docs with userId as current user and convert it to an array of postIds **(likedPosts)** since userId will be same in all. 
    - Liked status for UI can be checked using `likedPosts.includes(postId)`.
    - Like/Unlike would be pushing/splicing from likedPost array.
    - This approach would require 2 API calls: updating count in document of posts collection and creating/deleting document in likes collection.

Please suggest which approach to implement, or if you'd like to suggest any other approach.

**Comment by vinayak-vohra** [comment]
@Sanchitbajaj02 I've implemented the like feature, but it requires some changes in the structure of likes collection. So, I need to discuss a few things related to this. Please reply when you are available.


https://github.com/Sanchitbajaj02/palettegram/assets/80148777/dd5b0ec3-b450-474b-a685-5346b87a413d

**Comment by Sanchitbajaj02** [comment]
> @Sanchitbajaj02 I've implemented the like feature, but it requires some changes in the structure of likes collection. So, I need to discuss a few things related to this. Please reply when you are available.

My availability is and will be a bit low. I know it needs to be structured. You can DM me on discord about your idea and we can discuss over there
